plugin-server/src/worker/plugins/run.ts

```diff
@@ -1,99 +1,74 @@
-import { PluginEvent, ProcessedPluginEvent } from '@posthog/plugin-scaffold'
+import { PluginEvent, ProcessedPluginEvent, RetryError } from '@posthog/plugin-scaffold'
 
-import { Hub, PluginConfig, PluginTaskType } from '../../types'
+import { Hub, PluginConfig, PluginTaskType, VMMethods } from '../../types'
 import { processError } from '../../utils/db/error'
-import { IllegalOperationError } from '../../utils/utils'
+import { delay, IllegalOperationError } from '../../utils/utils'
 import { Action } from './../../types'
 
-export async function runOnEvent(server: Hub, event: ProcessedPluginEvent): Promise<void> {
-    const pluginsToRun = getPluginsForTeam(server, event.team_id)
+export const ON_CAUSE_MAX_ATTEMPTS = 5
+export const ON_CAUSE_RETRY_MULTIPLIER = 2
+export const ON_CAUSE_RETRY_BASE_MS = 5000
+
+export async function runOnEvent(hub: Hub, event: ProcessedPluginEvent): Promise<void> {
```

Comment:
Q: This behavior isn't clear from code but what happens if this plugin server dies before we process the whole delay?

Related: In the non-split plugin server, this could lead to session timeouts/events being processed multiple times. Should we be voiding these promises and awaiting on them in the background instead of the critical path?

Reply:
Yeah, `onEvent` might run more than once for an event if the server goes down and the Kafka offset isn't committed, and this does elevate the risk of session timeouts. We can either:

1. Keep this as is – but yeah, then Kafka consumption might stall due to session timeouts, so not great
1. `void runOnEvent` – and risk events to be skipped when the plugin server/`runOnEvent` goes down, not great either
1. Send a consumer `heartbeat()` before each retry – this would avoid session timeouts, though gets me a bit worried about queue throughput
1. Add a new Kafka topic (`async_event_handler_retries`) and send to it with a `setTimeout()`. I wanted to avoid the infra complexity of adding a new topic, but in truth it would definitely be the most reliable way.

What do you think?

Reply:
I think a variant on (2) makes sense, agreed with issues around 3) and 4). Two suggestions:

- note that `hub` has a `PromiseManager` built in - completely voided promises are problematic but this or something similar could help
- we could run the first `onX` function sync and void the follow-ups

Also happy to bring this discussion sync.

Reply:
Note there's a 5th option which would be leveraging graphile jobs for retries. :thinking: